### PR TITLE
feat(loopback4-authorization): RFIT-188 Casbin integration

### PR DESCRIPTION
**Pull request by vineet-suri**
**What is implemented**:
1. Casbin authorisation is integrated within current interface. New implementation will accept the following in the decorator - permissions array, name of resource, and optional boolean 'isCasbinPolicy' to choose to use casbin policy format (default is false).
2. Casbin_enforcer_config_getter is implemented. The implementation of this provider on application side will give information to initialise casbin enforcer. It gives option to either provide permissions data in casbin policy format - as file or from database (isCasbinPolicy = true), or to pass the allowed resources for a given user and resource type( isCasbinPolicy = false).
3. Casbin_resouce_modifier_provider: This has to be implemented on the application side. This provider will be invoked in sequence.ts before the authorisation step. This provider takes parsed arguments from request as input, and modifies the resource value based on the route params for dynamic use case of authorisation. This provider returns the modified resource value, which is passed to authorisation method.

So, to use this - 
1. Add casbin config file.
2. modify decorator params.
3. Implement Casbin_enforcer_config_getter provider
    (a) If using default casbin format files, then provide the policy file in this provider.
    (b) In other case, just provide the allowed resources for given user.
4. Implement Casbin_resouce_modifier_provider, and provide modified resource value to authorisation provider.
5. Bind providers in application.ts
6. Add relevant provider methods in sequence.ts (Refer sample app below for reference)

Authorisation package will use these to give authorisation decision using casbin.

Here is code for sample application implementing the above changes in extension: 
https://github.com/vineet-suri/casbin_integration_sample_app
Decorator is implemented in ping controller.

### CODE REVIEW

File: src/keys.ts [hunk 1]
```diff
@@ -23,14 +38,31 @@ export namespace AuthorizationBindings {
     'sf.userAuthorization.actions.userPermissions',
   );
 
+  export const CASBIN_ENFORCER = BindingKey.create<CasbinEnforcerFn<string>>(
+    'sf.userCasbinAuthorization.casbinenforcer',
+  );
+
+  export const CASBIN_ENFORCER_CONFIG_GETTER = BindingKey.create<CasbinEnforcerConfigGetterFn>(
```
Comment (samarpan-b) [inline src/keys.ts:45]: use same prefix.
sf.userAuthorization.actions.casbin.config

File: src/keys.ts [hunk 2]
```diff
@@ -23,14 +38,31 @@ export namespace AuthorizationBindings {
...
+    'sf.userCasbinAuthorization.casbinEnforcerConfigGetter',
+  );
+
+  export const CASBIN_RESOURCE_MODIFIER_FN = BindingKey.create<CasbinResourceModifierFn>(
```
Comment (samarpan-b) [inline src/keys.ts:49]: use same prefix.
sf.userAuthorization.actions.casbin.resourceModifier

File: src/keys.ts [hunk 3]
```diff
@@ -23,14 +38,31 @@ export namespace AuthorizationBindings {
...
+  );
+
   export const CONFIG = BindingKey.create<AuthorizationConfig>(
     'sf.userAuthorization.config',
   );
 
   export const PATHS_TO_ALLOW_ALWAYS = 'sf.userAuthorization.allowAlways';
+
+
+  export const RESOURCE_ID = BindingKey.create<string>('sf.resourceId');
```
Comment (samarpan-b) [inline src/keys.ts:60]: use same prefix.
sf.userAuthorization.actions.casbin.resource

File: src/providers/casbin-authorization-action.provider.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  ) { }
+
+  value(): CasbinAuthorizeFn {
+    return (response, resource) => this.action(response, resource);
+  }
+
+  async action(user: IAuthUserWithPermissions, resource: string): Promise<boolean> {
+    let authDecision = false;
+    try {
+      const metadata: AuthorizationMetadata = await this.getCasbinMetadata();
```
Comment (samarpan-b) [inline src/providers/casbin-authorization-action.provider.ts:27]: run prettier

File: src/providers/casbin-authorization-action.provider.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+      const action = metadata.permissions && metadata.permissions.length > 0 ? metadata.permissions[0] : DEFAULT_SCOPE;
+
+      const fn = await this.getCasbinEnforcerConfig();
+
+      const result = await fn(user, metadata.resource);
+
+      let enforcer: casbin.Enforcer;
+
+      if (metadata.isCasbinPolicy) {
```
Comment (samarpan-b) [inline src/providers/casbin-authorization-action.provider.ts:41]: can u put some documentation of what's going around here ?

File: src/types.ts [hunk 6]
```diff
@@ -10,13 +13,28 @@ export interface AuthorizeFn {
...
 }
 
+/**
+ * Authorize action method interface
+ */
+export interface CasbinAuthorizeFn {
+  // userPermissions - Array of permission keys granted to the user
+  // This is actually a union of permissions picked up based on role
+  // attached to the user and allowed permissions at specific user level
+  (user: IAuthUserWithPermissions, resVal: string): Promise<boolean>;
```
Comment (samarpan-b) [inline src/types.ts:23]: update the documentation here

File: src/types.ts [hunk 7]
```diff
@@ -63,3 +81,21 @@ export interface UserPermission<T> {
 export interface UserPermissionsFn<T> {
   (userPermissions: UserPermission<T>[], rolePermissions: T[]): T[];
 }
+
+export interface CasbinEnforcerFn<T> {
+  (model: T, policy: T | PostgresAdapter): Promise<casbin.Enforcer>;
```
Comment (samarpan-b) [inline src/types.ts:86]: add docs

**Review by samarpan-b — CHANGES_REQUESTED** [review]
Please update readme with usage details and see if u can add some test cases for new implementation

File: src/component.ts [hunk 8]
```diff
@@ -1,37 +1,44 @@
...
-import {AuthorizeActionProvider} from './providers/authorization-action.provider';
-import {AuthorizationMetadataProvider} from './providers/authorization-metadata.provider';
-import {UserPermissionsProvider} from './providers/user-permissions.provider';
-import {AuthorizationConfig} from './types';
 
+@bind({ tags: { [ContextTags.KEY]: AuthorizationBindings.COMPONENT.key } })
 export class AuthorizationComponent implements Component {
   providers?: ProviderMap;
-  bindings?: Binding[];
+  bindings1?: Binding[];
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply src/component.ts:14]: reverted

File: package.json [hunk 9]
```diff
@@ -51,13 +51,16 @@
     "@loopback/rest": "^1.16.3"
   },
   "dependencies": {
+    "@sourceloop/core": "^1.0.0-alpha.9",
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply package.json:54]: we are implementing an empty provider in extension for getting casbin enforcer config from the application. App needs to implement this provider to give us information casbin config. So, the app would need some user interface in this method to provide the required information. That is why i included IAuthUserWithPermissions interface from sourceloop/core.

How can we offer a generic user interface from extension side, to be implemented in app?

File: package.json [hunk 10]
```diff
@@ -51,13 +51,16 @@
...
+    "casbin": "^5.1.3",
+    "casbin-pg-adapter": "^1.4.0",
     "lodash": "^4.17.15"
   },
   "devDependencies": {
     "@loopback/boot": "^1.4.4",
     "@loopback/build": "^2.0.3",
-    "@loopback/context": "^1.20.2",
-    "@loopback/core": "^1.8.5",
+    "@loopback/context": "^3.8.2",
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply package.json:62]: done

File: src/decorators/authorize.decorator.ts [hunk 11]
```diff
@@ -1,12 +1,14 @@
...
 
-export function authorize(permissions: string[]) {
+export function authorize(metadata: AuthorizationMetadata) {
   return MethodDecoratorFactory.createDecorator<AuthorizationMetadata>(
     AUTHORIZATION_METADATA_ACCESSOR,
     {
-      permissions: permissions || [],
+      permissions: metadata.permissions || [],
+      resource: metadata.resource || '',
+      isCasbinPolicy: metadata.isCasbinPolicy || false
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply src/decorators/authorize.decorator.ts:11]: This is metadata object used to create the decorator function.

File: src/types.ts [hunk 12]
```diff
@@ -10,13 +13,28 @@ export interface AuthorizeFn {
...
+}
 /**
  * Authorization metadata interface for the method decorator
  */
 export interface AuthorizationMetadata {
   // Array of permissions required at the method level.
   // User need to have at least one of these to access the API method.
   permissions: string[];
+
+  resource: string;
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply src/types.ts:33]: added documentation for all new interfaces.

File: src/keys.ts [hunk 13]
```diff
@@ -23,14 +38,31 @@ export namespace AuthorizationBindings {
     'sf.userAuthorization.actions.userPermissions',
   );
 
+  export const CASBIN_ENFORCER = BindingKey.create<CasbinEnforcerFn<string>>(
+    'sf.userCasbinAuthorization.casbinenforcer',
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply src/keys.ts:42]: removed

File: src/keys.ts (hunk 1 above)
Comment (vineet-suri) [reply src/keys.ts:45]: done

File: src/keys.ts (hunk 2 above)
Comment (vineet-suri) [reply src/keys.ts:49]: done

File: src/providers/casbin-authorization-action.provider.ts (hunk 4 above)
Comment (vineet-suri) [reply src/providers/casbin-authorization-action.provider.ts:27]: done

File: src/providers/casbin-authorization-action.provider.ts (hunk 5 above)
Comment (vineet-suri) [reply src/providers/casbin-authorization-action.provider.ts:41]: done

File: src/types.ts (hunk 6 above)
Comment (vineet-suri) [reply src/types.ts:23]: done

File: src/types.ts (hunk 7 above)
Comment (vineet-suri) [reply src/types.ts:86]: done

File: src/keys.ts (hunk 3 above)
Comment (vineet-suri) [reply src/keys.ts:60]: done

File: src/keys.ts [hunk 14]
```diff
@@ -1,20 +1,35 @@
...
   AuthorizationMetadata,
   UserPermissionsFn,
   AuthorizationConfig,
+  CasbinEnforcerFn,
+  CasbinAuthorizeFn,
+  CasbinEnforcerConfigGetterFn,
+  CasbinResourceModifierFn,
 } from './types';
+import { CoreBindings } from '@loopback/core';
+import { AuthorizationComponent } from './component';
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply src/keys.ts]: removed

File: src/keys.ts [hunk 15]
```diff
@@ -1,20 +1,35 @@
...
 
 /**
  * Binding keys used by this component.
  */
 export namespace AuthorizationBindings {
+
+  export const COMPONENT = BindingKey.create<AuthorizationComponent>(
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply src/keys.ts]: removed

File: src/component.ts [hunk 16]
```diff
@@ -1,37 +1,44 @@
...
+import { UserPermissionsProvider } from './providers/user-permissions.provider';
+import { AuthorizationConfig } from './types';
 
-import {AuthorizationBindings} from './keys';
-import {AuthorizeActionProvider} from './providers/authorization-action.provider';
-import {AuthorizationMetadataProvider} from './providers/authorization-metadata.provider';
-import {UserPermissionsProvider} from './providers/user-permissions.provider';
-import {AuthorizationConfig} from './types';
 
+@bind({ tags: { [ContextTags.KEY]: AuthorizationBindings.COMPONENT.key } })
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply src/component.ts]: removed

File: src/policy.csv [hunk 17]
```diff
@@ -0,0 +1,2 @@
+p, u0851e3b-156b-4b7f-85e9-48f0953a6cc8, session, CreateMeetingSession
```
Comment (samarpan-b) [inline src/policy.csv:1]: can we remove this ? If not, can we keep it as something generic,like CreateUser

File: src/types.ts [hunk 18]
```diff
@@ -1,4 +1,5 @@
-import {Request} from '@loopback/rest';
+import {Request} from '@loopback/express';
+import PostgresAdapter from 'casbin-pg-adapter';
```
Comment (samarpan-b) [inline src/types.ts:2]: why  import like this ? Dont we need {} ?

File: src/types.ts [hunk 19]
```diff
@@ -10,13 +11,33 @@ export interface AuthorizeFn {
...
   // User need to have at least one of these to access the API method.
   permissions: string[];
+
+  // Name of resource for which authorisation is being sought
+  resource?: string;
+
+  /**
+   * Boolean flag to determine whether we are using casbin policy format or not
+   * isCasbinPolicy = true, when we are providing casbin format policy doc in application
+   * isCasbinPolicy = false, when we are impplementing provider in app to give casbin policy
```
Comment (samarpan-b) [inline src/types.ts:38]: implementing - typo

File: src/types.ts [hunk 20]
```diff
@@ -63,3 +84,71 @@ export interface UserPermission<T> {
...
+  identifier?: ID;
+  permissions: string[];
+  authClientId: number;
+  userPreferences?: IUserPrefs;
+  email?: string;
+  role: string;
+  firstName: string;
+  lastName: string;
+  middleName?: string;
+  tenantId?: TID;
```
Comment (samarpan-b) [inline src/types.ts:151]: can we eliminate tenantid, usertenantid and password expiry from here ? That's too app specific

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: README.md [hunk 21]
```diff
@@ -37,13 +41,21 @@ For a quick starter guide, you can refer to our [loopback 4 starter](https://git
...
-- Add component to application.
+- Add component to application. Also add providers to implement casbin authorisation.
 
 ```ts
 this.bind(AuthorizationBindings.CONFIG).to({
   allowAlwaysPaths: ['/explorer'],
 });
 this.component(AuthorizationComponent);
+
+this.bind(AuthorizationBindings.CASBIN_ENFORCER_CONFIG_GETTER).toProvider(
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply README.md:52]: No, this is not optional. Purpose of this provider is to provide casbin model definition, and policy data, for creation of casbin enforcer in the extension. There are 2 ways to provide policy, but either have to be provided through this provider only.

Other provider is also required, although we can the resource parameter unaltered through that.

File: README.md [hunk 22]
```diff
@@ -21,6 +21,10 @@ It provides three ways of integration
 2. **Role based permissions** - Permissions are associated to roles and users have a specific role attached. This actually reduces redundancy in DB a lot, as most of the time, users will have many common permissions. If that is not the case for you, then, use method #1 above.
 3. **Role based permissions with user level override** - This is the most flexible architecture. In this case, method #2 is implemented as is. On top of it, we also add user-level permissions override, allow/deny permissions over role permissions. So, say there is user who can perform all admin role actions except he cannot remove users from the system. So, DeleteUser permission can be denied at user level and role can be set as Admin for the user.
 
+As a further enhancement to these methods, we are using [casbin library!](https://casbin.org/docs/en/overview) to define permissions at level of entity or resource associated with an API call. Casbin authorisation implementation can be performed in two ways:
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply README.md:24]: Done

File: README.md [hunk 23]
```diff
@@ -37,13 +41,21 @@ For a quick starter guide, you can refer to our [loopback 4 starter](https://git
 
 In order to use this component into your LoopBack application, please follow below steps.
 
-- Add component to application.
+- Add component to application. Also add providers to implement casbin authorisation.
```
[the remark this answers is not in the document]
Comment (vineet-suri) [reply README.md:44]: Ok

File: src/policy.csv (hunk 17 above)
Comment (vineet-suri) [reply src/policy.csv:1]: This file is used to write policy for the current request. So, it gets re-written every time a request is made.

So, just emptying it for now.

File: src/types.ts (hunk 18 above)
Comment (vineet-suri) [reply src/types.ts:2]: PostgresAdapter is a default export from the casbin-pg-adapter library. That is why such a syntax.

File: src/types.ts (hunk 19 above)
Comment (vineet-suri) [reply src/types.ts:38]: corrected

File: src/types.ts (hunk 20 above)
Comment (vineet-suri) [reply src/types.ts:151]: done

**Review by samarpan-b — APPROVED** [review]

**Pull request merged**